contrib/vcl/source/vcl_io_handle.cc

```diff
@@ -0,0 +1,768 @@
+    // must come from a dual stack socket. An actual v6 address can come from
+    // both dual stack socket and v6 only socket. If |peer_addr| is an actual v6
+    // address and the socket is actually v6 only, the returned address will be
+    // regarded as a v6 address from dual stack socket. However, this address is not going to be
+    // used to create socket. Wrong knowledge of dual stack support won't hurt.
+    return *Envoy::Network::Address::addressFromSockAddr(addr, len, /*v6only=*/false);
+  } catch (const EnvoyException& e) {
+    PANIC(fmt::format("Invalid remote address for fd: {}, error: {}", sh, e.what()));
+  }
+}
+
+void vclEndptFromAddress(vppcom_endpt_t& endpt,
+                         Envoy::Network::Address::InstanceConstSharedPtr address) {
+  endpt.is_cut_thru = 0;
+  if (address->ip()->version() == Envoy::Network::Address::IpVersion::v4) {
+    const sockaddr_in* in = reinterpret_cast<const sockaddr_in*>(address->sockAddr());
+    endpt.is_ip4 = 1;
+    endpt.ip = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(&in->sin_addr));
+    endpt.port = static_cast<uint16_t>(in->sin_port);
+  } else {
+    const sockaddr_in6* in6 = reinterpret_cast<const sockaddr_in6*>(address->sockAddr());
+    endpt.is_ip4 = 0;
+    endpt.ip = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(&in6->sin6_addr));
+    endpt.port = static_cast<uint16_t>(in6->sin6_port);
+  }
+}
+
+// Converts a VCL return types to IoCallUint64Result.
+Api::IoCallUint64Result vclCallResultToIoCallResult(const int32_t result) {
+  if (result >= 0) {
+    // Return nullptr as IoError upon success.
+    return Api::IoCallUint64Result(
+        result, Api::IoErrorPtr(nullptr, Envoy::Network::IoSocketError::deleteIoError));
+  }
+  RELEASE_ASSERT(result != VPPCOM_EINVAL, "Invalid argument passed in.");
+  return Api::IoCallUint64Result(
+      /*rc=*/0, (result == VPPCOM_EAGAIN
+                     // EAGAIN is frequent enough that its memory allocation should be avoided.
+                     ? Api::IoErrorPtr(Envoy::Network::IoSocketError::getIoSocketEagainInstance(),
+                                       Envoy::Network::IoSocketError::deleteIoError)
+                     : Api::IoErrorPtr(new Envoy::Network::IoSocketError(-result),
+                                       Envoy::Network::IoSocketError::deleteIoError)));
+}
+
+} // namespace
+
+VclIoHandle::~VclIoHandle() {
+  if (VCL_SH_VALID(sh_)) {
+    VclIoHandle::close();
+  }
+}
+
+Api::IoCallUint64Result VclIoHandle::close() {
+  VCL_LOG("closing sh {:x}", sh_);
+  RELEASE_ASSERT(VCL_SH_VALID(sh_), "sh must be valid");
+  int rc = 0, wrk_index;
+
+  wrk_index = vclWrkIndexOrRegister();
+
+  if (is_listener_) {
+    if (wrk_index) {
+      uint32_t sh = wrk_listener_->sh();
+      RELEASE_ASSERT(wrk_index == vppcom_session_worker(sh), "listener close on wrong thread");
+      clearChildWrkListener();
+      rc = vppcom_session_close(sh);
```

Comment:
After the changes in the vcl_event yes it is. We're not invalidating sh_ because we're waiting for the destructor to be called on main thread, at which point vppcom_session_close can also be called. Comment added.